### ISSUE TITLE
Name '_' is not defined with gettext.install()

### ISSUE THREAD

**Issue by Dreamsorcerer** [python/mypy#8727]
I get `Name '_' is not defined` after using `gettext.install()`. Is there some configuration option to specify this is a builtin?

**Comment by msullivan** [comment]
I don't think so. It might be possible to do with a plugin?

**Comment by JukkaL** [comment]
Modifying the stubs of builtins in typeshed locally and using `--custom-typeshed-dir` is perhaps the easiest way to make this work.

Better support for monkeypatching stdlib might be a reasonable feature to have at some point, if we can implement it in a simple enough way. One option would to write a generic plugin for this.

**Comment by Dreamsorcerer** [comment]
Is there any way to atleast ignore any `name-defined` errors globally? Valid errors are caught by basic flake8 linting, so I don't think mypy is actually adding anything for me with these error codes.

**Comment by aivarannamaa** [comment]
Pylint has [additional-builtins](http://pylint.pycqa.org/en/stable/technical_reference/features.html#variables-checker-options) configuration option. It would be great if MyPy had something like this.

**Comment by Dreamsorcerer** [comment]
I can spend a full week working on Mypy shortly, is this something I'm likely to be able to do? Is adding a config option to extend the builtins an acceptable solution? If so, can you point me to the code I would need to look at. Thanks.

**Comment by nicoddemus** [comment]
> Is adding a config option to extend the builtins an acceptable solution?

Can't speak for the maintainers, but an option that allows me to point to an `.pyi` file with "extra builtins" would solve this problem for me.

**Comment by JelleZijlstra** [comment]
A way that would work right now is to fork typeshed to add `_` to builtins, then pass `--custom-typeshed-dir` to your mypy invocation. Doing it without `--custom-typeshed-dir` is considerably harder; I'm guessing you'll have to find where mypy internally retrieves the `builtins` stub and inject your new builtin.

**Comment by JelleZijlstra** [comment]
(Also, I'm not sure a PR adding the ability to inject builtins would be accepted.)

**Comment by nicoddemus** [comment]
> A way that would work right now is to fork typeshed to add _ to builtins, then pass --custom-typeshed-dir to your mypy invocation. 

Indeed that's what I did and I can confirm it works.

> (Also, I'm not sure a PR adding the ability to inject builtins would be accepted.)

Indeed.

**Comment by Dreamsorcerer** [comment]
> A way that would work right now is to fork typeshed to add `_` to builtins, then pass `--custom-typeshed-dir` to your mypy invocation.

But, that requires cloning and maintaining the typeshed repo, right? That doesn't feel like the right solution to close this ticket. I'm expecting something that can extend Mypy's behaviour without requiring that extra overhead.

e.g. Following the suggestion to have a .pyi file that can be used to extend the builtiins; this would only require maintaining a 1 line file to support my example use case.

**Comment by JelleZijlstra** [comment]
> But, that requires cloning and maintaining the typeshed repo, right? That doesn't feel like the right solution to close this ticket. I'm expecting something that can extend Mypy's behaviour without requiring that extra overhead.

Yes, it definitely has some overhead, but on the plus side it doesn't require deep changes to mypy.

Another hack that I think would work is to copy just `builtins.pyi` from typeshed, add your custom globals, and put it somewhere in your `MYPYPATH`. That means you have to maintain a fork of just one file instead of all of mypy.

**Comment by Dreamsorcerer** [comment]
Seems the latter is not an option:
```mypy: "stubs/builtins.pyi" shadows library module "builtins"
note: A user-defined top-level module with name "builtins" is not supported```

**Comment by nicoddemus** [comment]
> Another hack that I think would work is to copy just builtins.pyi from typeshed, add your custom globals, and put it somewhere in your MYPYPATH. That means you have to maintain a fork of just one file instead of all of mypy.

That unfortunately does not work it seems:

```
```

**Comment by huyz** [comment]
Another package that `install()`s into `builtins` and for which a simple solution would be useful: https://github.com/gruns/icecream

**Comment by matthewrv** [comment]
> > A way that would work right now is to fork typeshed to add _ to builtins, then pass --custom-typeshed-dir to your mypy invocation.
> 
> Indeed that's what I did and I can confirm it works.
> 
> > (Also, I'm not sure a PR adding the ability to inject builtins would be accepted.)
> 
> Indeed.

Sorry, maybe I misunderstood something. I tried to fork typeshed, modify builtins.pyi and add it to mypy but i got error with this traceback. Am I doing something wrong?

```
(.venv) D:\>mypy test.py --custom-typeshed-dir D:\my_project\typeshed --show-traceback
test.py:1: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.812
Traceback (most recent call last):
  File "C:\Users\matthewrv\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\matthewrv\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "mypy\semanal.py", line 4835, in accept
  File "mypy\nodes.py", line 1030, in accept
  File "mypy\semanal.py", line 3306, in visit_expression_stmt
  File "mypy\nodes.py", line 1555, in accept
  File "mypy\semanal.py", line 3581, in visit_call_expr
  File "mypy\nodes.py", line 1487, in accept
  File "mypy\semanal.py", line 3509, in visit_name_expr
  File "mypy\semanal.py", line 4023, in lookup
IndexError: string index out of range
test.py:1: : note: use --pdb to drop into pdb
```

As I undestrand this happens here:
https://github.com/python/mypy/blob/d089891198ef470c8bec9bd7d7b50a02757c5b68/mypy/semanal.py#L4023-L4025

**Comment by Dreamsorcerer** [comment]
See https://github.com/python/mypy/issues/9802

The person you quoted was using another function (which does not start with _).

**Comment by Rosadosa** [comment]
Any update on implementing a way to ignore the _() that gets installed by gettext?
Would be great to have a builtin tags like flake8 and pylint have!

**Comment by Dreamsorcerer** [comment]
My impression from various comments is that this practice is discouraged (even if the gettext documentation has not been updated to say this...), and it's unlikely to be supported any time soon (if ever).

I'm fine with closing this issue as wont-fix if any of the mypy developers want to confirm that is the case, rather than just leaving it open as a vague hope that it might happen.
